File: src/forecast_staging.py
"""Write candidate forecasts to a LOCAL staging file (isolated from Supabase).

The maintainer inner-loop tests a promoted candidate on the local app *before*
deploying. `run_daily_forecast.py --staging` writes the forecast rows here instead
of upserting to Supabase; the backend serves them when `HEATWAVE_FORECAST_FILE`
is set. JSON shape matches what `src/routes/forecast.ts` returns (lat/lon merged
for the map, ISO date strings) so the frontend needs no changes.
"""
from __future__ import annotations

import datetime as _dt
import json
import os


def _iso(v) -> str:
    """date / datetime / Timestamp / str -> ISO-8601 string (midnight for dates)."""
    if isinstance(v, str):
        return v
    if isinstance(v, _dt.datetime):
        return v.isoformat()
    if isinstance(v, _dt.date):
        return _dt.datetime(v.year, v.month, v.day).isoformat()
    # pandas Timestamp or anything date-like
    iso = getattr(v, "isoformat", None)
    return iso() if callable(iso) else str(v)


def to_staging_records(rows, provinces) -> list[dict]:
    """Forecast row dicts + a provinces frame (id/lat/lon) -> JSON-able records
    carrying lat/lon and ISO dates (one record per province x horizon)."""
    latlon = {int(p.id): (float(p.lat), float(p.lon)) for p in provinces.itertuples()}
    out = []
    for r in rows:
        pid = int(r["province_id"])
        lat, lon = latlon.get(pid, (None, None))
        sw = r.get("swbgt_pred")
        out.append({
            "province_id": pid,
            "lat": lat,
            "lon": lon,
            "target_date": _iso(r["target_date"]),
            "generated_at": _iso(r["generated_at"]),
            "horizon_days": int(r["horizon_days"]),
            "probability": float(r["probability"]),
            "predicted_label": bool(r["predicted_label"]),
            "swbgt_pred": None if sw is None else float(sw),
            "risk_level": r["risk_level"],
            "model_version": r["model_version"],
        })
    return out


def write_staging(rows, provinces, path: str) -> int:
    """Write staging records to ``path`` (JSON array). Returns the row count."""
    records = to_staging_records(rows, provinces)
    os.makedirs(os.path.dirname(path) or ".", exist_ok=True)
    with open(path, "w", encoding="utf-8") as f:
        json.dump(records, f)
    return len(records)

File: src/forecast_staging.ts
import { readFileSync } from "fs";

/**
 * Local staging forecast source — mirrors the `heatwave.forecasts` queries in
 * src/routes/forecast.ts but reads from a JSON file written by
 * `run_daily_forecast.py --staging`. Lets the backend serve a promoted candidate's
 * forecasts for a pre-deploy local test WITHOUT touching Supabase. Active only when
 * `HEATWAVE_FORECAST_FILE` is set.
 */
export interface StagingRow {
  province_id: number;
  lat: number | null;
  lon: number | null;
  target_date: string;
  generated_at: string;
  horizon_days: number;
  probability: number;
  predicted_label: boolean;
  swbgt_pred: number | null;
  risk_level: string;
  model_version?: string;
}

const dateOnly = (s: string): string => String(s).slice(0, 10);

export function readStaging(path: string): StagingRow[] {
  const data = JSON.parse(readFileSync(path, "utf-8"));
  if (!Array.isArray(data)) throw new Error("staging file is not a JSON array");
  return data as StagingRow[];
}

/**
 * One row per province for the map: from the freshest `generated_at` run, the
 * soonest upcoming `target_date` (>= today). Mirrors the DISTINCT ON query.
 */
export function stagingMap(rows: StagingRow[], today: string): StagingRow[] {
  const best = new Map<number, StagingRow>();
  for (const r of rows) {
    if (dateOnly(r.target_date) < today) continue; // only upcoming
    const cur = best.get(r.province_id);
    if (
      !cur ||
      r.generated_at > cur.generated_at ||
      (r.generated_at === cur.generated_at && r.target_date < cur.target_date)
    ) {
      best.set(r.province_id, r);
    }
  }
  return [...best.values()];
}

/** Latest batch for one province: freshest generated_at, upcoming dates, up to `days`. */
export function stagingProvince(
  rows: StagingRow[], provinceId: number, days: number, today: string,
): StagingRow[] {
  const mine = rows.filter((r) => r.province_id === provinceId);
  if (mine.length === 0) return [];
  const maxGen = mine.reduce((m, r) => (r.generated_at > m ? r.generated_at : m), "");
  return mine
    .filter((r) => r.generated_at === maxGen && dateOnly(r.target_date) >= today)
    .sort((a, b) => (a.target_date < b.target_date ? -1 : 1))
    .slice(0, days);
}
